gcc/rust/expand/rust-attribute-visitor.h

```diff
@@ -40,19 +40,70 @@ class AttrVisitor : public AST::ASTVisitor
   void expand_trait_function_decl (AST::TraitFunctionDecl &decl);
   void expand_trait_method_decl (AST::TraitMethodDecl &decl);
 
-  template <typename T> void expand_pointer_allow_strip (T &values)
+  /**
+   * Expand a set of values, erasing them if they are marked for strip, and
+   * replacing them with expanded macro nodes if necessary.
+   * This function is slightly different from `expand_pointer_allow_strip` as
+   * it can only be called in certain expansion contexts - where macro
+   * invocations are allowed.
+   *
+   * @param ctx Context to use for macro expansion
+   * @param values Iterable reference over values to replace or erase
+   * @param extractor Function to call when replacing values with the content
+   * 		of an expanded AST node
+   */
+  template <typename T, typename U>
+  void expand_macro_children (MacroExpander::ContextType ctx, T &values,
+			      std::function<U (AST::SingleASTNode)> extractor)
   {
+    expander.push_context (ctx);
+
     for (auto it = values.begin (); it != values.end ();)
```

Comment:
is there a way we could rewrite this for loop so it always continues to the  end even if we miss incremenentations in the body?

If you hit a case in the fort loop where you would want the iteratator not to move forward you could always issue it--;